src/collections/vec.rs

```diff
@@ -1806,35 +1837,79 @@ impl<'bump, T: 'bump + Copy> Vec<'bump, T> {
+        // * Source and destination data ranges cannot overlap as we just reserved the destination
+        //   range from the bump.
+
+        unsafe {
+            self.extend_from_slice_copy_unchecked(other);
+            // The elements at `old_len..new_len` are initialized by `extend_from_slice_copy_unchecked`
+            // above. Update the Vec's length to reflect the number of items added.
+            self.set_len(new_len);
+        }
+    }
+
+    /// For each slice in `slices`, copies all elements in the slice and appends them to the `Vec`.
+    ///
+    /// This method is equivalent to calling [`extend_from_slice_copy`] in a loop, but is able
+    /// to precompute the total amount of space to reserve in advance. This reduces the potential
+    /// maximum number of reallocations needed from one-per-slice to just one.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bumpalo::{Bump, collections::Vec};
+    ///
+    /// let b = Bump::new();
+    ///
+    /// let mut vec = bumpalo::vec![in &b; 1];
+    /// vec.extend_from_slices_copy(&[&[2, 3], &[], &[4]]);
+    /// assert_eq!(vec, [1, 2, 3, 4]);
+    /// ```
+    ///
+    /// ```
+    /// use bumpalo::{Bump, collections::Vec};
+    ///
+    /// let b = Bump::new();
+    ///
+    /// let mut vec = bumpalo::vec![in &b; 'H' as u8];
+    /// vec.extend_from_slices_copy(&["ello,".as_bytes(), &[], " world!".as_bytes()]);
+    /// assert_eq!(vec, "Hello, world!".as_bytes());
+    /// ```
+    ///
+    /// [`extend_from_slice_copy`]: #method.extend_from_slice_copy
+    pub fn extend_from_slices_copy(&mut self, slices: &[&[T]]) {
+        // Reserve the total amount of capacity we'll need to safely append the aggregated contents
+        // of each slice in `slices`.
+        let bytes_to_reserve: usize = slices.iter().map(|buf| buf.len()).sum();
+        self.reserve(bytes_to_reserve);
+
+        // SAFETY:
         // * `dst` is valid for writes of `other.len()` bytes as `self.reserve(other.len())`
         //   above guarantees that.
-        // * Because `src` is a `&[T]` and dst is a `&[T]` within the `Vec<T>`,
-        //   `copy_nonoverlapping`'s alignment requirements are met.
         // * Source and destination ranges cannot overlap as we just reserved the destination
         //   range from the bump.
         unsafe {
-            let src = other.as_ptr();
-            let dst = self.as_mut_ptr().add(old_len);
-            ptr::copy_nonoverlapping(src, dst, other.len());
+            // Copy the contents of each slice onto the end of `self`
+            slices.iter().for_each(|slice| {
+                self.extend_from_slice_copy_unchecked(slice);
+                // Update the Vec's length to reflect the number of items we just added. This needs
+                // to happen for each slice so the next call to `extend_from_slice_copy_unchecked`
+                // will write to the correct (updated) offset.
+                self.set_len(self.len() + slice.len());
```

Comment:
I guess all callers of `extend_from_slice_copy_unchecked` are required to update lengths themselves. Can we document that in the method's docs as another thing that is the caller's responsibility?